[Sema] Fix compiler error when extending a typealias of a partially specialized generic type

xavgru12 opened the pull request:
Resolves #68212.
Relax the fall-through type alias criteria a little bit by allowing a mismatch in the number of generic parameters at the greatest depth, in case that difference is due to some of those generic parameters being bound to fully concrete types.

xavgru12 committed:
comment out isPassThroughTypealies

xavgru12 committed:
comment out isPassThroughTypealies

AnthonyLatsis commented:
@swift-ci please smoke test Linux

saehejkang commented:
i was curious to see how this was solved and guess I was overcomplicating issue 😭

AnthonyLatsis commented:
This is far from solved, FWIW 🙂. I only triggered the tests to encourage the author to iterate on failures before any feedback arrives.

xedin commented:
Is it possible to make this PR as a draft?

AnthonyLatsis commented:
For posterity, the convert to draft button is at the bottom of the **Reviewers** section.

xavgru12 committed:
merge with main

xavgru12 committed:
copy isPassThroughTypealias to isTypeInferredByTypealias and comment unneeded checks

xavgru12 committed:
copy isPassThroughTypealias to isTypeInferredByTypealias and comment unneeded checks

xavgru12 committed:
add declaration of new function

xavgru12 committed:
add declaration of new function

xavgru12 committed:
merge with main

xavgru12 committed:
merge with main

xavgru12 committed:
merge with main

xavgru12 committed:
checkout main

xavgru12 committed:
remove oudtdated file

xavgru12 committed:
first version

xavgru12 committed:
returns only true if no unsupported type conversion is entered

xavgru12 committed:
add back spaces

xavgru12 committed:
remove more unwanted changes

xavgru12 committed:
get rid of unneeded changes

xavgru12 committed:
according to github issue, this is supposed to fail, but the error message shall be different

xavgru12 committed:
add unit test for the new supported case

xavgru12 commented:
I forgot that I am not able to start the CI lol

xavgru12 committed:
remove unneeded include

xavgru12 committed:
commit forced blank lines

xavgru12 committed:
remove unneeded copy

xavgru12 committed:
remove extra brackets

xavgru12 committed:
remove comment

xavgru12 committed:
format code

xavgru12 committed:
change variable name

xavgru12 committed:
change location of function

xavgru12 committed:
remove unneeded check§

xavgru12 committed:
correct location

xavgru12 committed:
Revert "commit forced blank lines"

This reverts commit 2a534871dd13682ab7835fc9171187c864d38b03.

xavgru12 committed:
Revert "commit forced blank lines"

This reverts commit 2a534871dd13682ab7835fc9171187c864d38b03.

xavgru12 committed:
Revert "commit forced blank lines"

This reverts commit 2a534871dd13682ab7835fc9171187c864d38b03.

xavgru12 committed:
Revert "commit forced blank lines"

This reverts commit 2a534871dd13682ab7835fc9171187c864d38b03.

xavgru12 committed:
revert evaluate ternary operator

xavgru12 committed:
whitespace

xavgru12 committed:
whitespace

xavgru12 committed:
add missing bracket

xavgru12 commented:
@AnthonyLatsis check this out

xavgru12 committed:
Merge branch 'main' into merge_in_typealias

xavgru12 commented:
merged main again

xavgru12 commented:
@AnthonyLatsis I see you have been quite busy with 6.1 these days. Could you review this?

AnthonyLatsis commented:
@swift-ci please smoke test macOS

AnthonyLatsis commented:
Will have a review ready by Monday.

AnthonyLatsis commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3055,6 +3058,38 @@ bool TypeChecker::isPassThroughTypealias(TypeAliasDecl *typealias,
+  bool isInferredType = false;
+  auto nominalGenericArguments = nominal->getDeclaredInterfaceType()
+                                     .getPointer()
+                                     ->getAs<BoundGenericType>()
```
This cast will fail if `nominal` does not have a generic parameter list (`nominal->isGeneric()`). The name is confusing, I know. `nominal->getGenericSignature()` is a different check:

```swift
// Has generic parameter list; generic signature '<T>'
struct S1<T> {
  // No generic parameter list; generic signature '<T>'
  struct S2 {}

  // Has generic parameter list; generic signature <T, U where T == Int>
  struct S3<U> where T == Int {}
}
```

xavgru12 replied:
if (!nominal->isGeneric())
return false
before I do the cast,
since the function would not make any sense anyway. Is this fine?

AnthonyLatsis replied:
Sounds good. Might as well `ASSERT(nominal->isGeneric() == typealias->isGeneric())` after that.

xavgru12 replied:
I removed the assert with ad5d532 since it breaks 4 tests:
Compatibility/stdlib_generic_typealiases.swift, decl/nested/type_in_type.swift, stdlib/Filter.swift, stdlib/Map.swift

I do have:
```
  if (nominal->isGeneric() != typealias->isGeneric())
    return false;
```

Let me know if this is fine

AnthonyLatsis commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3055,6 +3058,38 @@ bool TypeChecker::isPassThroughTypealias(TypeAliasDecl *typealias,
   return false;
 }
 
+bool TypeChecker::isTypeInferredByTypealias(TypeAliasDecl *typealias,
```
Can be a static function.

AnthonyLatsis commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3032,11 +3032,14 @@ bool TypeChecker::isPassThroughTypealias(TypeAliasDecl *typealias,
+
+  // Check for inferred types.
   if (nominalGenericParams.size() != typealiasGenericParams.size())
-    return false;
+    return isTypeInferredByTypealias(typealias, nominal);
+
+  // Check that the type parameters are the same the whole way through.
   if (!std::equal(nominalGenericParams.begin(), nominalGenericParams.end(),
                   typealiasGenericParams.begin(),
                   [](GenericTypeParamType *gp1, GenericTypeParamType *gp2) {
```
This condition evaluates to true only if there is an equal number of generic parameters at each depth. For example, the following type alias does not qualify:

```swift
struct S1<A, B> {
  // Generic parameters: <A, B, C>
  // Depth:               0  0  1
  struct Nested<C> {}
}

struct S2<A> {
  // Generic parameters: <A, B, C>
  // Depth:               0  1  1
  typealias Alias<B, C> = S1<A, B>.Nested<C>
}
```

We should keep exercising this rule for all depths except the last.

xavgru12 replied:
Is this an expected failing or passing example? Seems like A may be specialized depending on the argument. How would I iterate over the depths as in getNextDepth() to check for the generic arguments size?

AnthonyLatsis replied:
Yes, this should still fail the "pass through" check. Could be made to work, but I don’t think that’s a reasonable goal for this PR. It’s going to require changes to generic requirement inference for extensions, and I’m not quite sure if the language steering group would be fine to take in that sort of generalization without a public discussion and/or proposal.
___

The generic parameter types returned by `GenericSignature::getGenericParams` are ordered by increasing depth. Here is what we can do:
1. `if (nominalGenericParams.size() != typealiasGenericParams.size())`
   1. Obtain the max depth across both arrays of generic parameter types.
   1. Remove elements at this depth from both arrays.
   1. If their sizes still don’t match, give up as we currently do. If they do, proceed to perform the `std::equal` check.
2. Return `isTypeInferredByTypealias` if the check succeeds and the original sizes don’t match.

xavgru12 replied:
You are saying to do the std::equal check before the isTypeInferredByTypealias check. 
"Return isTypeInferredByTypealias if the check succeeds"
What do you mean by check here?

xavgru12 replied:
Thanks for the clarification. I am working on it. Hopefully the other changes from code review look fine for you.

xavgru12 replied:
> We need to perform this check regardless of whether the original number of generic params matched.

This is what you said 4 days ago. However what I took from this thread is to do the std::equal check before `isTypeInferredByTypealias`. Confused here, since I am not supposed to do std::equal check after. So if I put it before since it is supposed to be regardless of whether the original number of generic params matched, it will return false before I am starting the if clause for `isTypeInferredByTypealias`. Please help me out here.

AnthonyLatsis reviewed (commented):
Overall direction is looking great!
___

Please add regression tests with extensions for all the type alias examples from my comments as you go.

xavgru12 committed:
make the new function static

xavgru12 committed:
add typealiasBoundGenericType->hasTypeParameter()

xavgru12 committed:
add test case for typealiasBoundGenericType->hasTypeParameter() inferred specialized nested types

xavgru12 committed:
remove redundant check

xavgru12 committed:
fix typo

xavgru12 committed:
breaking expected fail

xavgru12 commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3055,6 +3058,38 @@ bool TypeChecker::isPassThroughTypealias(TypeAliasDecl *typealias,
+    if (nominalBoundGenericType.getPointer()->getKind() ==
+        typealiasBoundGenericType.getPointer()->getKind()) {
+      continue;
+    }
```
I tried nominalBoundGenericType.getPointer()->isEqual(typealiasBoundGenericType.getPointer())
and the test case where I put the expected fail succeeded again(however it is supposed to fail). So it definitely would make it more strict. Would this be a good solution?

xavgru12 committed:
add test case for making type check stricter

xavgru12 committed:
add isGeneric check

xavgru12 committed:
add assert

xavgru12 committed:
make sure no exotic cases slip through

xavgru12 committed:
add test case for expected fail

xavgru12 committed:
add test case extension for expected pass

xavgru12 commented:
@AnthonyLatsis 
I changed the code according review and added test cases. Let me know what you think about this.

xavgru12 commented:
@AnthonyLatsis reminder

AnthonyLatsis reviewed (commented):
Thank you for the reminders and the patience. Another somewhat random test case I think worth having:

```swift
struct S1<T> {
  struct Inner<U> {}
}

struct S2<T> {
  typealias A1<U> = S1<T>.Inner<U>
  typealias A2<U> = S1<T>.Inner<U> where T == Int
}

extension S2<Int>.A1 {
  func foo1() {
    let int: Int
    let _: T = int // OK
  }
}

extension S2.A2 {
  func foo2() {
    let int: Int
    let _: T = int // OK
  }
}
```

xavgru12 committed:
Merge remote-tracking branch 'origin/main' into generic_typealias

xavgru12 committed:
rename struct to fix redeclaration after merge

xavgru12 committed:
remove assert since it breaks 4 tests: Compatibility/stdlib_generic_typealiases.swift, decl/nested/type_in_type.swift, stdlib/Filter.swift, stdlib/Map.swift

xavgru12 committed:
remove superfluous getPointer

xavgru12 committed:
add another test case

xavgru12 committed:
remove isInferredType

xavgru12 committed:
minimize diagnostic noise

xavgru12 committed:
add test case

xavgru12 committed:
changed foreach to while

xavgru12 committed:
update test case with where

xavgru12 committed:
remove comments

xavgru12 commented on test/decl/ext/specialize.swift:
```diff
@@ -32,6 +32,66 @@ extension IntFoo where U == Int {
+  func adding(_ c: Int) -> Self {
+        S4.Nested(c: self.c + c) //expected-error {{binary operator '+' cannot be applied to operands of type 'C' and 'Int'}} expected-note {{overloads for '+' exist with these partially matching parameter lists: (Int, Int)}}
+    }
+}
```
Thanks!

xavgru12 commented on test/decl/ext/specialize.swift:
```diff
@@ -32,6 +32,66 @@ extension IntFoo where U == Int {
+  func adding(_ x: Int) -> Self {
+    S2(x: self.x + x, y: y, z: z) // expected-error {{binary operator '+' cannot be applied to operands of type 'X' and 'Int'}} expected-note {{overloads for '+' exist with these partially matching parameter lists: (Int, Int)}}
+  }
+}
```
Thanks!

xavgru12 commented:
@AnthonyLatsis 
I iterated over it and added some comments unresolving threads. Please review again and I hope you are able to keep the overview over this, so let me know if something is unclear.

xavgru12 commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3032,17 +3069,41 @@ bool TypeChecker::isPassThroughTypealias(TypeAliasDecl *typealias,
+      return false;
+    }
+
+    if (!std::equal(nominalGenericParams.begin(), nominalGenericParams.end(),
+                    typealiasGenericParams.begin(),
+                    [](GenericTypeParamType *gp1, GenericTypeParamType *gp2) {
+                      return gp1->isEqual(gp2);
+                    })) {
+      return false;
+    }
```
Confused here, added a comment above about this

AnthonyLatsis replied:
The comment: https://github.com/swiftlang/swift/pull/73169#discussion_r2129591844
___

So the goal is to relax the fall-through type alias criteria a little bit by allowing a mismatch in the number of generic parameters at the greatest depth, but *only* if that difference is due to some of those generic parameters being bound to fully concrete types. As for other depths, the current rules still apply. For example, `<A><B,C>...` vs. `<A, B><C>...` or `<A, B, C>...` vs. `<A, B>...` must not meet the criteria regardless of the nature of `...`. I.e., we should still return early if the size check or the `std::equal` check fail for what remains after we shave off max depth elements.

Could you rebase (I have renamed `isGeneric` to `hasGenericParamList`) and see if the tests pass with the current change? I would expect some of them to fail. If not, we should get this case covered.

xavgru12 commented:
@AnthonyLatsis another reminder

AnthonyLatsis commented:
@slavapestov Could you take a look at this too?

xavgru12 commented:
@AnthonyLatsis @slavapestov 
Could someone have a look at it, so I can iterate further on it if required?

xavgru12 commented:
@AnthonyLatsis

xavgru12 committed:
Merge branch 'main' into generic_typealias

xavgru12 committed:
use new method name

xavgru12 committed:
new variant
swift::TypeBase::getKind() const /home/xaver/code/swift-project2/swift/include/swift/AST/Types.h:595:73

Failed Tests (5):
  Swift(linux-x86_64) :: AssociatedTypeInference/associated_type_tuple.swift
  Swift(linux-x86_64) :: Generics/requirement_inference.swift
  Swift(linux-x86_64) :: Generics/tuple-conformances.swift
  Swift(linux-x86_64) :: IRGen/tuple_conformances.swift
  Swift(linux-x86_64) :: SILOptimizer/tuple-conformances.swift

xavgru12 committed:
make sure tuple tests pass

xavgru12 committed:
return true if function is true

xavgru12 committed:
rename function

xavgru12 committed:
simplify nullptr check

xavgru12 commented:
I am getting closer to what you have in mind.
I resolved the failing tuple tests.

About the remaining crashing test in Generics/requirement_inference.swift

GIven:
```
protocol P1 { 
  func p1()
}

protocol P2 : P1 { }

struct X1<T : P1> { 
  func getT() -> T { }
}

struct X5<T : P2> { }


typealias X1WithP2MoreArgs<T: P2, U> = X1<T>


extension X1WithP2MoreArgs {
  func bad2() {
    _ = X5<T>() // expected-error{{type 'T' does not conform to protocol 'P2'}}
  }
}
```
The test crashes. With my changes isPassThroughTypealias returns true, instead of previously false.
The reason is that 
```
nominalGenericParams.size() != typealiasGenericParams.size()

xavgru12 committed:
generic requirements test case passes

xavgru12 committed:
nested test case passes

xavgru12 committed:
adjust test case

xavgru12 committed:
remove some extra brackets

xavgru12 committed:
clang format file

xavgru12 commented:
@AnthonyLatsis it would be great to review it, so it can be merged after 6.4 is branched off?

xavgru12 commented:
@xedin Could you take a look at this too?
